Replace the rust-crypto crate to other crates

somniumism opened the pull request:
The `rust-crypto@2.3.6` crate that we are using now is unaudited and has not received a commit since 2016. We are still unknown whether there are any vulnerabilities offhand. But if there are ones, they will never receive fixes. Thus, we think that it is necessary to completely replace the rust-crypto crate to other crypto crates.

I replaced hash in the `rust-crypto` crate to other crates(`sha-1`, `sha2`, `sha3`, `ripemd160` and `digest`), and blake2 in the `rust-crypto` crate to `blake2` crate. And I'm working on replacing `aes` and `block mode` to other crates(`block-modes`, `aes`, `aes-soft`).

I think there are unnecessary and dirty codes, so I need your review to modify them.

Commit:
Add /target directory to .gitignore

Commit:
Replace hash crates to other crates

Commit:
Replace the blake crate to other crate

Commit:
Replace crates related to the 256-CBC and 128-CTR mode to other crates

sgkim126 commented:
The first and second commit looks good to be merged.
How about splitting the PR?

sgkim126 commented on src/blake.rs:
```diff
@@ -56,17 +56,19 @@ macro_rules! implement_blake {
                 let input = s.as_ref();
                 let mut result = Self::default();
-                let mut hasher = Blake2b::new(result.len());
+                let mut hasher = VarBlake2b::new(result.len()).unwrap();
```
How about it?
```rust
macro_rules! implement_blake {
    ($self:ident, $size:expr) => {
        impl Blake for $self {
            fn blake<T: AsRef<[u8]>>(s: T) -> Self {
                Self::blake_with_key(s, &[])
            }
            fn blake_with_key<T: AsRef<[u8]>>(s: T, key: &[u8]) -> Self {
                let input = s.as_ref();
                let mut hasher = VarBlake2b::new_keyed(&key, $size);
                hasher.input(input);
                let mut result: [u8; $size] = [0; $size];
                result.copy_from_slice(&hasher.vec_result());
                Self(result)
            }
        }
    };
}

implement_blake!(H128, 16);
implement_blake!(H160, 20);
implement_blake!(H256, 32);
implement_blake!(H512, 64);
```
It will bypass the clippy error.

somniumism replied:
Sounds good! To solve this error, I replaced fixed-hash@0.2.5 to @0.2.2, but this methods looks better. I'll modify it. Thank you.

somniumism commented:
@sgkim126 For the time being, I'm waiting for @HoOngEe's review. And I think it's fine to split the PR. By the way, if there is a special reason for doing it, could you tell me the reason?

sgkim126 commented:
The 4th commit seems to take more time to be implemented.
If you can implement it before long, I think splitting is not necessary.

somniumism commented:
I'll try to complete the 4th commit within today. If I can't finish it by today, I'll split it up.

HoOngEe commented:
It looks good to me up to the 3rd commit. I'll review the fourth when you finish it.

somniumism commented:
@sgkim126 @HoOngEe I replaced crates related to aes-256-cbc to other crates. I'm waiting for your review. : ) 

There are still two parts left to be fixed: 1) ctr mode, and 2) error handling.
To solve 1) ctr mode, I used the [ctr crate](https://docs.rs/ctr/0.3.2/ctr/). However, there seems to be no way to make encryption and decoding by dividing them into two functions. This crate is designed to allow decryption only if the object `cipher = Aes128Ctr::new(&key, &nonce)` is shared. The object can be created in the encryption process, and when divided into two functions, it is difficult to convey the object to the decryption function. Unfortunately, existing crates may still need to be used. How about it?

And I don't know how to solve 2) error handling. To be exact, I don't understand our error handling structure, and roles of `ScryptError`, `ring::error::Unspecified` and `SymmetricCipherError`. So, I need your advice.

If the third and forth commit looks good to be merged too, I'll split it up.

sgkim126 commented on src/aes.rs:
```diff
@@ -14,59 +14,35 @@
-    let mut buffer = [0; 4096];
-    let mut write_buffer = RefWriteBuffer::new(&mut buffer);
-
+    let cipher = Aes256Cbc::new_var(&key, &iv.to_be_bytes()).unwrap();
```
```rust
let cipher = Aes256Cbc::new_var(&key, &iv.to_be_bytes()).map_err(SymmetricCipherError::InvalidLength)?;
```

sgkim126 replied:
Or change the type of the result to `Result<Vec<u8>, Result<Vec<u8>, InvalidKeyIvLength>>` after re-exporting `block_modes::InvalidKeyIvLength`.

seonpyoKim replied:
Let's change the result type and remove all the rcrypto (rust-crypto renamed) dependencies

sgkim126 commented on src/aes.rs:
```diff
@@ -14,59 +14,35 @@
-        finish = is_underflow(decryptor.decrypt(&mut read_buffer, &mut write_buffer, true)?);
-        final_result.extend(write_buffer.take_read_buffer().take_remaining().iter().cloned());
-    }
+    let cipher = Aes256Cbc::new_var(&key, &iv.to_be_bytes()).unwrap();
```
ditto

sgkim126 reviewed (changes requested):
It seems that there is no test to check whether the encrypted result has been changed.
Please add tests like https://github.com/CodeChain-io/rust-codechain-crypto/blob/0e2cc259dbc0c50a67ad267d8d42ad940345bf29/src/blake.rs#L108.

HoOngEe commented:
I checked the functions in `ctr` crate. We can implement encryption and decryption using `apply_keystream` method in `SyncStreamCipher` trait. However I still cannot understand why ctr does not provide `StreamCipher` methods. I think it still remains unimplemented.

HoOngEe reviewed (commented):
I checked the functions in ctr crate. We can implement encryption and decryption using apply_keystream method in SyncStreamCipher trait. However I still cannot understand why ctr does not provide StreamCipher methods. I think it still remains unimplemented.

Commit:
Replace crates related to  scrypt to other crates

somniumism commented on src/scrypt.rs:
```diff
@@ -32,8 +32,8 @@ pub fn derive_key(pass: &Password, salt: &[u8; 32], n: u32, p: u32, r: u32) -> R
-    let scrypt_params = ScryptParams::new(log_n, r, p);
-    scrypt(pass.as_bytes(), salt, &scrypt_params, &mut derived_key);
+    let scrypt_params = ScryptParams::new(log_n, r, p).unwrap();
+    let _ = scrypt(pass.as_bytes(), salt, &scrypt_params, &mut derived_key);
```
@sgkim126 I need your help. This code need error handling because the output format of `scrypt()` is `Result<(), InvalidOutputLen>`. However, I'm not used to doing error handling yet. I didn’t know whether to use `map_err()` or `unwrap_or_else()` and how to use these, so I wrote _ for the time being.

seonpyoKim replied:
You need to implement type conversion function from the foreign type to the local type. If conversion is possible, you can use `?` operator simply. I'll help you for the problem.

somniumism commented on Cargo.toml:
```diff
@@ -7,8 +7,18 @@ edition = "2018"
 [dependencies]
 ring = "0.14.6"
 quick-error = "1.2"
-rust-crypto = "0.2.36"
```
I removed completely the `rust-crypto` crate.

somniumism commented on src/aes.rs:
```diff
@@ -146,4 +125,32 @@ mod tests {
     }
+
+    #[test]
+    fn aes_256_cbc_encrypt_decrypt() {
```
I added tests to check to check whether the encrypted result has been changed.

somniumism commented:
@sgkim126 @HoOngEe I replaced crates related to the `aes-128-ctr` mode to other crates, and I made a commit combined two commits, replacing the `256-cbc` mode and replacing the `128-ctr` mode. I need your review : )

I'm working on replacing `scrypt` mode to other crates, and I have [a problem](https://github.com/CodeChain-io/rust-codechain-crypto/pull/8#discussion_r375066937).  The function `scrypt()`'s output format is `Result<(), InvalidOutputLen>`, so it seems it needs error handling. but I don't know how to solve it.

somniumism commented:
I replaced all modules in the `rust-crypto` crate to other crates. I removed completely the `rust-crypto` crate from `Crago.toml`. Therefore, from now on, we don't use the `rust-crypto` crate. 

@sgkim126 @HoOngEe Could you review this PR?

seonpyoKim commented on src/scrypt.rs:
```diff
@@ -32,8 +32,8 @@ pub fn derive_key(pass: &Password, salt: &[u8; 32], n: u32, p: u32, r: u32) -> R
     let mut derived_key = vec![0u8; KEY_LENGTH];
-    let scrypt_params = ScryptParams::new(log_n, r, p);
-    scrypt(pass.as_bytes(), salt, &scrypt_params, &mut derived_key);
+    let scrypt_params = ScryptParams::new(log_n, r, p).unwrap();
```
You should also handle the error related to parameters

somniumism replied:
I modified it. Thank you : )

sgkim126 commented on src/scrypt.rs:
```diff
@@ -32,8 +32,8 @@ pub fn derive_key(pass: &Password, salt: &[u8; 32], n: u32, p: u32, r: u32) -> R
     let mut derived_key = vec![0u8; KEY_LENGTH];
-    let scrypt_params = ScryptParams::new(log_n, r, p);
-    scrypt(pass.as_bytes(), salt, &scrypt_params, &mut derived_key);
+    let scrypt_params = ScryptParams::new(log_n, r, p)?;
```
IMO, this change needs a test to check whether the result is the same as the previous implementation.

somniumism replied:
I added a [test](https://github.com/CodeChain-io/rust-codechain-crypto/pull/8/commits/a687270eaca77140177d4cd8683f725919518448#r375666923).

somniumism commented on src/scrypt.rs:
```diff
@@ -32,9 +32,43 @@ pub fn derive_key(pass: &Password, salt: &[u8; 32], n: u32, p: u32, r: u32) -> R
+    use crate::password::Password;
+
+    #[test]
+    fn scrypt_test() {
```
@sgkim126 How about it? I added a test to check whether the result is the same as the previous implementation. Expected values, `right_bits` and `left_bits` were brought from the previous implementation. : )

sgkim126 replied:
LGTM

HoOngEe reviewed (approved):
LGTM
